### ISSUE TITLE
hermes should not create IBC clients and updates when threshold >= 1

### ISSUE THREAD

**Issue by ancazamfir** [informalsystems/ibc-rs#1149]
## Crate
relayer-cli, relayer

## Summary of Bug
hermes updates clients:
- as part of message relaying, 
- refresh of clients used by opened channels (to prevent client expiration), 
- when misbehavior is detected and it requires intermediate headers to be sent in client updates,
- when `hermes update client ` CLI is used

It is currently possible to configure clients with threshold == 1 or > 1.
- example with threshold > 1  (`trust_threshold = { numerator = '6', denominator = '5' }`):
```
Success: ChainError(
    "deliver_tx for Tx hash D2147251A6818BC5B0C6ABF22444575B571751C30F0EECCB4C20186BD3CCADFE reports error: code=Err(1), log=Log(\"internal\")",
)
```
- when threshold == 1 (`trust_threshold = { numerator = '6', denominator = '5' }`) the create client succeeds but any client update will result in sequential verification and consensus states for all headers being installed on chain. Maybe we should even consider setting some max threshold to something like `2/3`. 

Hermes should not use/ update clients with threshold == 1 or in cases where the number of updates is bigger than some threshold. At least not by default. We could offer on override for the `hermes update client` CLI:
 - another problem when trying the CLI is websocket connection reset. In the example below the client update would require 90k headers to be pulled, verified and Tx-ed with `MsgUpdateClient`. But we never get there as the CLI fails with:
 ```
$ hermes update client ibc-1 07-tendermint-2
    Finished dev [unoptimized + debuginfo] target(s) in 2.10s
     Running `target/debug/hermes update client ibc-1 07-tendermint-2`
Jul 05 08:11:30.084  INFO using default configuration from '/Users/ancaz/.hermes/config.toml'
Jul 05 08:11:30.182 TRACE subscribing to query: tm.event = 'Tx' chain.id=ibc-1
Jul 05 08:11:30.188 TRACE subscribing to query: tm.event = 'NewBlock' chain.id=ibc-1
Jul 05 08:11:30.189 TRACE subscribed to all queries chain.id=ibc-1
Jul 05 08:11:30.190 DEBUG starting event monitor chain.id=ibc-1
Jul 05 08:11:30.203 TRACE subscribing to query: tm.event = 'Tx' chain.id=ibc-2
Jul 05 08:11:30.204 TRACE subscribing to query: tm.event = 'NewBlock' chain.id=ibc-2
Jul 05 08:11:30.205 TRACE subscribed to all queries chain.id=ibc-2
Jul 05 08:11:30.205 DEBUG starting event monitor chain.id=ibc-2
Jul 05 08:11:30.227 TRACE light client verification trusted=2-3385 target=2-92769
Jul 05 08:20:48.167 ERROR failed to collect events: WebSocket driver failed: WebSocket Error: failed to read from WebSocket connection: WebSocket protocol error: Connection reset without closing handshake (code: 1) chain.id=ibc-2
Jul 05 08:20:48.169 TRACE trying to reconnect to WebSocket endpoint ws://127.0.0.1:26457/websocket chain.id=ibc-2
Jul 05 08:20:48.171 TRACE error when reconnecting: failed to create WebSocket driver: WebSocket Error: IO error: Connection refused (os error 61) (code: 1) chain.id=ibc-2
```
Creating a client with threshold 1, connection and channel is permissionless and a very cheap attack on hermes if there is no filtering: hermes ends up performing a lot of client updates that quickly drain the relayer wallets.

When fixing this issue we should also not allow threshold < 1/3. Currently when trying to create a client with < 1/3 (in the example below `trust_threshold = { numerator = '1', denominator = '5' }`) the chain responds with error:

```
Success: ChainError(
    "deliver_tx for Tx hash E8FC6386C6A3E926DD571C5B199B1D13F0F82B20BB705EF82A364A117A8BF666 reports error: code=Err(1), log=Log(\"internal\")",
)
```

## Acceptance Criteria

- [ ] add validation to the config.toml trust threshold: any total trust threshold smaller than 1/3 should be rejected
- [ ] additionally, any trust threshold bigger or equal to 1 should also be rejected (numerator >= denom)
- [ ] warn that the `trust_threshold` from the config.toml is an "advanced" feature

____

## For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate milestone (priority) applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by romac** [comment]
Should we just enforce the trust threshold to be 1/3, and remove it from the configuration? Is there an actual use case for setting the trust threshold to something different? (aside from exercising various light client behaviors when testing)

Quoting @milosevic:

> [The trust threshold] should always be at least +1/3, as we assume there are at most 1/3 of faulty voting power. Using different threshold is a bit harder to justify from the protocol perspective; it can only be interpreted as more security, but in terms of intersection property it does not really bring anything.

**Comment by ancazamfir** [comment]
> Should we just enforce the trust threshold to be 1/3, and remove it from the configuration? Is there an actual use case for setting the trust threshold to something different? (aside from exercising various light client behaviors when testing)
> 
> Quoting @milosevic:
> 
> > [The trust threshold] should always be at least +1/3, as we assume there are at most 1/3 of faulty voting power. Using different threshold is a bit harder to justify from the protocol perspective; it can only be interpreted as more security, but in terms of intersection property it does not really bring anything.

We could put limits for the clients created by hermes. But anyone can create clients and chain allows everything in (1/3, 1] range. The question is should we refuse to update clients with 1? This would mean we don't relay packets for channels using that client, we don't refresh it etc.

**Comment by romac** [comment]
I suggest we stay very conservative and do the following:

a) remove the `trust_threshold` config option and always create client with a trust threshold of 1/3
b) refuse to update clients with a trust threshold different than 1/3

We can revisit this when someone comes to us with a compelling use case for using a trust threshold other than 1/3.

@ancazamfir @adizere @milosevic What do you think?

**Comment by adizere** [comment]
Notes from the sync call: Staying conservative is definitely a good idea. But a few concerns arose w.r.t. Solo Machine clients and exotic chains that may wish to use different trust threshold. In any case, Hermes has no control over that parameter, and it may be overly constraining. Taking into account also our idea to keep parameters documented in the config.toml directly, we decided on this acceptance criteria:

> Acceptance Criteria
> 
>   add validation to the config.toml trust threshold: any total trust threshold smaller or equal to 1/3 should be rejected
>   additionally, any trust threshold bigger or equal to 1 should also be rejected (numerator >= denom)
>   warn that the trust_threshold from the config.toml is an "advanced" feature

I guess we can still go for the purely conservative choice, and altogether remove this parameter and constrain Hermes to 1/3, but that require to make Hermes double-check the trust threshold upon constructing any ForeignClient. Since there is no "single responsibility" to creating Foreign clients in the codebase yet, this change would be a bigger one, so we can do it later on.

So to summarize:
1. as a first step, we do the minimal changes suggested in the acceptance criteria, and 
2. open an issue and plan to make Hermes more robst w.r.t. client trust threshold, by constraining that parameter to certain predefined values (1/3, but maybe additionally 2/3 also).

**Comment by hu55a1n1** [comment]
> any total trust threshold smaller or equal to 1/3 should be rejected

Shouldn't we allow trust thresholds equal to 1/3 because that is also the default?

**Comment by adizere** [comment]
> > any total trust threshold smaller or equal to 1/3 should be rejected
> 
> Shouldn't we allow trust thresholds equal to 1/3 because that is also the default?

Yes, thank you for catching that!
